llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,352 @@
+    HR.emitInScoreboard(Scoreboard, MI->getDesc(), MemoryBanks, MI->operands(),
+                        MI->getMF()->getRegInfo(), Cycle);
+
+    scheduleNode(SU, Actual);
+    LLVM_DEBUG(dbgs() << "Scoreboard\n"; Scoreboard.dumpFull(););
+  }
+  LLVM_DEBUG(dbgs() << "==== First iteration scheduled ======\n");
+  return true;
+}
+
+bool PostPipeliner::scheduleOtherIterations() {
+  // Make sure that all the copies can be placed at II from the previous one.
+  // This looks like overkill, but it accommodates dependences that span
+  // multiple loop edges. Without these, the pattern should repeat after the
+  // first set of copies.
+  int CurrentCycle = 0;
+  for (int L = NInstr; L < NTotalInstrs; L += NInstr) {
+    // Forget the previous iteration. We accumulate the resources sticking
+    // out of II, and by inserting all stages we will end up with
+    // the steady state resource allocation
+    for (int C = 0; C < II; C++) {
+      CurrentCycle++;
+      Scoreboard.advance();
+    }
+    for (int K = 0; K < NInstr; K++) {
+      const int N = L + K;
+      SUnit &SU = DAG->SUnits[N];
+      MachineInstr *MI = SU.getInstr();
+      // Earliest tracks the latencies of the loop carried deps
+      const int Earliest = Info[N].Earliest;
+      // Insert supplies the modulo condition.
+      const int Insert = Info[N - NInstr].Cycle + II;
+
+      // All iterations following the first one should fit exactly
+      LLVM_DEBUG(dbgs() << "Trying " << N << " at " << Insert
+                        << " CurrentCycle=" << CurrentCycle << "\n";
+                 Scoreboard.dumpFull());
+      if (Earliest > Insert) {
+        LLVM_DEBUG(dbgs() << "  Latency not met (Earliest=" << Earliest
+                          << ")\n");
+        return false;
+      }
+      if (fit(MI, Insert, 1, II) != Insert) {
+        LLVM_DEBUG(dbgs() << "  Resource conflict\n");
+        return false;
+      }
+      const MemoryBankBits MemoryBanks = HR.getMemoryBanks(MI);
+      const int LocalCycle = (Insert - CurrentCycle) % II;
+      LLVM_DEBUG(dbgs() << "  Emit in " << -Depth + LocalCycle << "\n");
+      HR.emitInScoreboard(Scoreboard, MI->getDesc(), MemoryBanks,
+                          MI->operands(), MI->getMF()->getRegInfo(),
+                          -Depth + LocalCycle);
+      scheduleNode(SU, Insert);
+    }
+  }
+  return true;
+}
+
+bool PostPipeliner::schedule(ScheduleDAGMI &TheDAG, int InitiationInterval) {
+  II = InitiationInterval;
+  DAG = &TheDAG;
+  NTotalInstrs = DAG->SUnits.size();
+  assert(NTotalInstrs % NInstr == 0);
+
+  // Let's not skimp on size here. This allows us to insert any instruction
+  // in the unrolled dag.
+  NCopies = NTotalInstrs / NInstr;
+  Depth = NCopies * II + HR.getPipelineDepth();
+  Scoreboard.reset(Depth);
+
+  Info.resize(NTotalInstrs);
+  LLVM_DEBUG(for (int I = 0; I < NInstr;
+                  I++) { dbgs() << I << " " << *DAG->SUnits[I].getInstr(); });
+  computeLoopCarriedEarliest();
```

Comment:
super-nit: The MachineScheduler does something similar and calls this `bias`. Maybe we could call this function `biasEarliesCyclesForLCDs`?

Reply:
I already renamed this, because it does more.